openlr/openlr_decoder.cpp

```diff
@@ -50,6 +69,142 @@ struct alignas(kCacheLineSize) Stats
   uint32_t m_tightOffsets = 0;
   uint32_t m_total = 0;
 };
+
+void ExpandFakes(Index const & index, Graph & g, Graph::EdgeVector & path)
+{
+  auto const isRealVertex = [&index](FeatureID const & fid, m2::PointD const & p) {
+    Index::FeaturesLoaderGuard g(index, fid.m_mwmId);
+    auto const ft = g.GetOriginalFeatureByIndex(fid.m_index);
+    bool matched = false;
+    ft->ForEachPoint(
+        [&p, &matched](m2::PointD const & fp) {
+          if (p == fp)
+            matched = true;
+        },
+        FeatureType::BEST_GEOMETRY);
+    return matched;
+  };
+
+  auto const expandFake = [&g, &path, isRealVertex](Graph::EdgeVector::iterator edgeIt) {
+    if (!edgeIt->IsFake())
+      return;
+
+    Graph::EdgeVector edges;
+    if (isRealVertex(edgeIt->GetFeatureId(), edgeIt->GetStartPoint()))
+      g.GetRegularOutgoingEdges(edgeIt->GetStartJunction(), edges);
+    else
+      g.GetRegularIngoingEdges(edgeIt->GetEndJunction(), edges);
+
+    CHECK(!edges.empty(), ());
+
+    auto const it = find_if(begin(edges), end(edges), [&edgeIt](Graph::Edge const & real) {
+      if (real.GetFeatureId() == edgeIt->GetFeatureId() && real.GetSegId() == edgeIt->GetSegId())
+        return true;
+      return false;
+    });
+
+    CHECK(it != end(edges), ());
+
+    if (2 * EdgeLength(*edgeIt) >= EdgeLength(*it))
+      *edgeIt = *it;
+    else
+      path.erase(edgeIt);
+
+    LOG(LDEBUG, ("len(e):", EdgeLength(*edgeIt), "len(*it):", EdgeLength(*it)));
+  };
+
+  expandFake(begin(path));
+  expandFake(--end(path));
+}
+
+template <typename InputIterator>
+// Returns an iterator pointing to the first edge that should not be cut off.
+InputIterator CutOffset(InputIterator start, InputIterator const stop, uint32_t const offset)
+{
+  if (offset == 0)
+    return start;
+  for (uint32_t distance = 0; start != stop; ++start)
+  {
+    auto const edgeLen = EdgeLength(*start);
+    if (distance <= offset && offset < distance + edgeLen)
+    {
+      // Throw out this edge if (offst - distance) is greater than edgeLength / 2.
+      if (2 * (offset - distance) >= edgeLen)
+        ++start;
+      break;
+    }
+    distance += edgeLen;
+  }
+
+  return start;
+}
+
+template <typename InputIterator, typename OutputIterator>
+void CopyWithoutOffsets(InputIterator const start, InputIterator const stop, OutputIterator out,
```

Comment:
s/CopyWithoutOffsets/CutOffsets/
?

Reply:
For me CutOffsets should not have an out parameter, it should perform it in-place.

Reply:
For example, function sort may do sort in-place (Lisp, C++), or purely (Haskell). So what? :)